kite-data/kite-data-core/src/main/java/org/kitesdk/data/PartitionExpression.java

```diff
@@ -58,8 +58,21 @@ public PartitionStrategy evaluate() {
        * FieldPartitioner. Instead, it thinks it's Object[].
        */
       List<FieldPartitioner> partitioners = Lists.newArrayList();
-      for (Object o : ((Object[]) object)) {
-        partitioners.add((FieldPartitioner) o);
+
+      Object[] objects = ((Object[]) object);
+
+      for (int i = 0; i < objects.length; i++) {
+
+        if (PartitionFunctions.IMMUTABLE == objects[i]) {
+
+          // adhere to the marker for immutable partitioners.
+          for (int j = 0; j < partitioners.size(); j++) {
+            partitioners.set(j, partitioners.get(j).asImmutable());
+          }
+
+        } else {
+          partitioners.add((FieldPartitioner) objects[i]);
+        }
```

Comment:
This supports the  immutable() function in the JEXL expression, which produces a sentinel value we check here. Note that previous Kite consumers won't be able to work with expressions that include immutable partitions, but I didn't see a way to add the immutability option to the expression in a way that would work for previous consumers.

I think this is okay, since users must opt into immutable partitions, and other partitions are unchanged.